### PR TITLE
Hyacinth 463 title typedef

### PR DESCRIPTION
**Pull request by barmintor**

### CODE REVIEW

**Comment by barmintor** [comment]
@elohanlon if we can get the initial language stuff merged I'll cherrypick the two commits added here into a new branch, and pick up the migration and front-end changes from there.

**Comment by barmintor** [comment]
I have rebased with updates for the refactor to `title.value` and `title.value_lang`. I have also broken the spec support refactors allowing `rspec --tag ~@\solr --tag ~@\fedora` to run into a separate commit (the first in the current PR, not tagged with HYACINTH-463). I think we may want to consider merging it separately (and first), as it should review quickly and it will clean the diff from the actual HYACINTH-463 work up a lot.

For reference:

- `rake hyacinth:ci` -> 1654 examples; 96.68% coverage
- `rspec --tag ~@\solr --tag ~@\fedora` -> 1355 examples; 90.45% coverage

**Comment by elohanlon** [comment]
@barmintor Ideally I'd like to implement a fast, in-memory search adapter at some later point so that we don't have to stub so much, and so that other tests that actually do rely on a real search adapter will run more quickly, but this is a good move for the tests you've modified.  Merging the commit separately sounds good too.  I've looked it over (https://github.com/cul/ldpd-hyacinth/pull/358/commits/8cb9d646a778cd62b76ac38aa2b09bdbbc031f72) and the only thing I caught was that you have one `solr: false` in there that I think should be omitted.

👍

**Comment by barmintor** [comment]
@elohanlon I know exactly which one you are talking about, and I just ran out of energy on the refactor TBQH. It might not be necessary since if I remember right the "sibling" test that does require solr has a `solr: true` flag.

**Comment by elohanlon** [comment]
@barmintor Given the position in the commit order, I'm fine with merging that commit as is and we can change the flag after this branch gets merged.  I'm all for keeping things simpler!

**Comment by barmintor** [comment]
ok - not sure why the commit is still here, it is marked as in main with the same hash, but you can safely select the other commits as a range and have less noise about changed files.

File: app/javascript/src/hyacinth_v1/components/digital_objects/metadata/Field.js
Comment (elohanlon) [inline app/javascript/src/hyacinth_v1/components/digital_objects/metadata/Field.js:13]: let uniqueFieldIdCounter = 0;

Comment (elohanlon) [inline app/javascript/src/hyacinth_v1/components/digital_objects/metadata/Field.js:13]:
```suggestion
let uniqueFieldIdCounter = 0;
```

File: app/javascript/src/hyacinth_v1/components/digital_objects/metadata/Field.js
Comment (elohanlon) [inline app/javascript/src/hyacinth_v1/components/digital_objects/metadata/Field.js:23]:
```suggestion
  sharedProps.inputName = sharedProps.inputName || `field-${uniqueCheckboxIdCounter++}`;
```
Two suggestions here:
1) Guaranteed unique id instead of a random one.
2) I didn't think `||=` works in js.  Was it working in your local copy?

Comment (elohanlon) [reply app/javascript/src/hyacinth_v1/components/digital_objects/metadata/Field.js:23]: Here's what threw me off:

```
let something ||= 'hello';
VM411:1 Uncaught SyntaxError: Unexpected token '||='
```
But:
```
let obj = {};
obj.something ||= 'hello';
"hello"
```

So I guess it only works for object properties.

Comment (elohanlon) [reply app/javascript/src/hyacinth_v1/components/digital_objects/metadata/Field.js:23]: fixed!

File: app/javascript/src/hyacinth_v1/components/digital_objects/rights/RightsEdit.js
Comment (elohanlon) [inline app/javascript/src/hyacinth_v1/components/digital_objects/rights/RightsEdit.js:36]:
```suggestion
  const { digitalObject, digitalObject: { digitalObjectType, title } } = digitalObjectData;
```
I forgot -- does this work?

Comment (barmintor) [reply app/javascript/src/hyacinth_v1/components/digital_objects/rights/RightsEdit.js:36]: I'm pretty sure nested destructures work, yes!

File: app/javascript/src/hyacinth_v1/components/digital_objects/rights/rights_form/subsections/DescriptiveMetadata.js
Comment (elohanlon) [inline app/javascript/src/hyacinth_v1/components/digital_objects/rights/rights_form/subsections/DescriptiveMetadata.js:90]:
```suggestion
          <ReadOnlyInput sm={8} inputName="rights_title_display" value={title?.value?.sortPortion} />
```
I think this will work now that we have the optional chaining babel addition.

File: app/models/digital_object.rb
Comment (elohanlon) [inline app/models/digital_object.rb:22]: It seems like these constants aren't doing a lot for us.  They're only used in the generate_label method below.  Maybe it's worth getting rid of them?  Or if it's useful to have globally-accessible constants like these we should put them in the title TypeDef class?

Also, `'value'` is a hard-coded string in the `generate_label` method, so if these keys are treated as special constants then maybe that should be too.

But I'm also fine with just getting rid of the constants and using strings in the generate_label method if that simplifies things.

File: app/models/digital_object.rb
Comment (elohanlon) [inline app/models/digital_object.rb:54]: I just realized that the downside of doing validations via typedef (rather than regular model validation) is that you can't provide a helpful message about what went wrong.

File: app/models/digital_object/asset.rb
Comment (elohanlon) [inline app/models/digital_object/asset.rb:32]:
```suggestion
  def generate_display_label
```
How would you feel about changing all occurrences of generate_label to generate_display_label?

File: lib/hyacinth/adapters/digital_object_search_adapter/solr/document_generator.rb
Comment (elohanlon) [inline lib/hyacinth/adapters/digital_object_search_adapter/solr/document_generator.rb:51]: Since label is often derived from the title, but not always, do you mind if we remove the display label stuff in here and add `'displayLabel_ss' => digital_object.generate_label` to the `merge_core_fields_for!` method?

Comment (elohanlon) [reply lib/hyacinth/adapters/digital_object_search_adapter/solr/document_generator.rb:51]: Actually, I just noticed that display_label is also added to the set of indexed titles.  Hmm, that does complicate things a bit.  Here's a question that we haven't answered yet: will people think of the display label AS the title when a title hasn't been assigned to the object?

File: lib/hyacinth/adapters/preservation_adapter/fedora3/dc_properties.rb
Comment (elohanlon) [inline lib/hyacinth/adapters/preservation_adapter/fedora3/dc_properties.rb:25]: Should we be getting the display label instead of the title, in case there is no title?

File: lib/hyacinth/adapters/preservation_adapter/fedora3/object_properties.rb
Comment (elohanlon) [inline lib/hyacinth/adapters/preservation_adapter/fedora3/object_properties.rb:11]: Same question here about using display_label instead of title.

File: lib/hyacinth/adapters/digital_object_search_adapter/solr/document_generator.rb
Comment (elohanlon) [reply lib/hyacinth/adapters/digital_object_search_adapter/solr/document_generator.rb:51]: After looking at some other parts of the code below, I'm starting to think that the display_label is more of a title than the (possibly absent) title is.

File: app/javascript/src/hyacinth_v1/components/shared/forms/inputs/ReadOnlyInput.js
Comment (barmintor) [reply app/javascript/src/hyacinth_v1/components/shared/forms/inputs/ReadOnlyInput.js:14]: I prefer to do this in another commit, because there is at least one other instance of this message in code that was not modified in this branch.

File: lib/hyacinth/adapters/digital_object_search_adapter/solr/document_generator.rb
Comment (elohanlon) [inline lib/hyacinth/adapters/digital_object_search_adapter/solr/document_generator.rb:52]:
```suggestion
          solr_document['display_label_ss'] = display_label
```
We've been going for snake case in solr keys to match the associated property names.

File: lib/hyacinth/adapters/digital_object_search_adapter/solr/document_generator.rb
Comment (elohanlon) [inline lib/hyacinth/adapters/digital_object_search_adapter/solr/document_generator.rb:56]:
```suggestion
            solr_document['title_sort_portion_ssi'] = digital_object.title.dig('value', 'sort_portion')
            solr_document['sort_title_ssi'] = digital_object.title.dig('value', 'sort_portion')
            solr_document['title_non_sort_portion_ssi'] = digital_object.title.dig('value', 'non_sort_portion')
```
Suggesting snake_case for key consistency.

File: app/models/digital_object.rb
Comment (barmintor) [reply app/models/digital_object.rb:54]: This is true. We could probably refactor to be better about it, but out-of-scope here.

File: app/models/digital_object/asset.rb
Comment (barmintor) [reply app/models/digital_object/asset.rb:32]: Done!

File: app/models/digital_object.rb
Comment (barmintor) [reply app/models/digital_object.rb]: Inlined!

File: app/javascript/src/hyacinth_v1/components/digital_objects/rights/rights_form/subsections/DescriptiveMetadata.js
Comment (barmintor) [reply app/javascript/src/hyacinth_v1/components/digital_objects/rights/rights_form/subsections/DescriptiveMetadata.js]: Done!

File: app/javascript/src/hyacinth_v1/components/digital_objects/rights/RightsEdit.js
Comment (barmintor) [reply app/javascript/src/hyacinth_v1/components/digital_objects/rights/RightsEdit.js]: Done!

File: app/javascript/src/hyacinth_v1/components/digital_objects/metadata/Field.js
Comment (barmintor) [reply app/javascript/src/hyacinth_v1/components/digital_objects/metadata/Field.js]: Done!

File: app/javascript/src/hyacinth_v1/components/digital_objects/metadata/Field.js
Comment (barmintor) [reply app/javascript/src/hyacinth_v1/components/digital_objects/metadata/Field.js]: Done!

**Review by elohanlon — CHANGES_REQUESTED** [review]
@barmintor Done with the review and this is great!  Thank you for all of your work on this.  I left some comments/suggestions.

File: lib/hyacinth/adapters/digital_object_search_adapter/solr/document_generator.rb
Comment (barmintor) [reply lib/hyacinth/adapters/digital_object_search_adapter/solr/document_generator.rb:56]: It's not tested (!), but Hyacinth::DigitalObject::SolrKeys actually camelcases the string keys (as does the Rights indexing) - this ensures no unexpected collisions here. If we want to revisit that,, we should ticket separately. I'll add a test for the SolrKeys behavior, though!

File: app/javascript/src/hyacinth_v1/components/shared/forms/inputs/ReadOnlyInput.js
Comment (barmintor) [reply app/javascript/src/hyacinth_v1/components/shared/forms/inputs/ReadOnlyInput.js]: Done!

File: app/models/digital_object.rb
Comment (elohanlon) [reply app/models/digital_object.rb:50]: 👍

File: lib/hyacinth/adapters/digital_object_search_adapter/solr/document_generator.rb
Comment (elohanlon) [reply lib/hyacinth/adapters/digital_object_search_adapter/solr/document_generator.rb]: Oh, you're right!  Comment withdrawn for now then!  Maybe worth revisiting later.

Comment (elohanlon) [reply lib/hyacinth/adapters/digital_object_search_adapter/solr/document_generator.rb]: And thank you for adding the test!

**Review by elohanlon — APPROVED** [review]
Thank you for the changes!

**Pull request merged**
